File: app/src/main/java/com/arukai/uajpspeak/adapter/MyRecyclerViewAdapter.java
package com.arukai.uajpspeak.adapter;

import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.TextView;

import com.arukai.uajpspeak.R;
import com.arukai.uajpspeak.model.Abecadlo;
import com.arukai.uajpspeak.model.DataObject;

import java.util.ArrayList;

import static com.arukai.uajpspeak.activity.HomeFragment.gender;

public class MyRecyclerViewAdapter extends RecyclerView
        .Adapter<MyRecyclerViewAdapter
        .DataObjectHolder> implements Filterable {
    private static String LOG_TAG = "MyRecyclerViewAdapter";
    private ArrayList<DataObject> mDataset;
    private static MyClickListener myClickListener;

    ArrayList<DataObject> mFilteredDataset;
    private UserFilter userFilter;
    private static Abecadlo abc;

    public static class DataObjectHolder extends RecyclerView.ViewHolder
            implements View
            .OnClickListener {
        TextView firstRow;
        TextView secondRow;
        TextView thirdRow;

        public DataObjectHolder(View itemView) {
            super(itemView);
            firstRow = itemView.findViewById(R.id.textView);
            secondRow = itemView.findViewById(R.id.textView2);
            thirdRow = itemView.findViewById(R.id.textView3);
            itemView.setOnClickListener(this);
            abc = new Abecadlo();
        }

        @Override
        public void onClick(View v) {
            v.setOnClickListener(null);
            myClickListener.onItemClick(getAdapterPosition(), v);
        }
    }

    public void setOnItemClickListener(MyClickListener myClickListener) {
        MyRecyclerViewAdapter.myClickListener = myClickListener;
    }

    public MyRecyclerViewAdapter(ArrayList<DataObject> myDataset) {
        mDataset = myDataset;
        mFilteredDataset = myDataset;
    }

    @Override
    public DataObjectHolder onCreateViewHolder(ViewGroup parent,
                                               int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.card_view_row, parent, false);

        return new DataObjectHolder(view);
    }

    @Override
    public void onBindViewHolder(DataObjectHolder holder, int position) {
        String part1 = mDataset.get(position).getmText1();
        String part2 = mDataset.get(position).getmText2();
        String part3 = mDataset.get(position).getmText3();

        String jp = "";
        String ukr = "";
        String phonetic = "";

        char code = part1.charAt(part1.length() - 1);
        if(code == gender || (code != 'm' && code != 'f')) {
            jp = part2;
            phonetic = abc.convert(part3);
            ukr = part3.replace("*", "");
        }

        holder.firstRow.setText(jp);
        holder.secondRow.setText(ukr);
        holder.thirdRow.setText(phonetic);
    }

    public void addItem(DataObject dataObj, int index) {
        mDataset.add(index, dataObj);
        notifyItemInserted(index);
    }

    public void deleteItem(int index) {
        mDataset.remove(index);
        notifyItemRemoved(index);
    }

    @Override
    public int getItemCount() {
        return mDataset.size();
    }

    public interface MyClickListener {
        void onItemClick(int position, View v);
    }

    @Override
    public Filter getFilter() {
        if(userFilter == null)
            userFilter = new UserFilter(this, mDataset);
        return userFilter;
    }

    private static class UserFilter extends Filter {

        private final MyRecyclerViewAdapter adapter;

        private final ArrayList<DataObject> originalList;

        private final ArrayList<DataObject> filteredList;

        private UserFilter(MyRecyclerViewAdapter adapter, ArrayList<DataObject> originalList) {
            super();
            this.adapter = adapter;
            this.originalList = new ArrayList<>(originalList);
            this.filteredList = new ArrayList<>();
        }

        @Override
        protected FilterResults performFiltering(CharSequence constraint) {
            filteredList.clear();
            final FilterResults results = new FilterResults();

            if (constraint.length() == 0) {
                filteredList.addAll(originalList);
            } else {
                final String filterPattern = constraint.toString().toLowerCase().trim();

                for (final DataObject user : originalList) {
                    String jpn = user.getmText2();
                    String ukr = user.getmText3().toLowerCase().replace("*", "");
                    if (jpn.contains(filterPattern) || ukr.contains(filterPattern)) {
                        filteredList.add(user);
                    }
                }
            }
            results.values = filteredList;
            results.count = filteredList.size();
            return results;
        }

        @Override
        protected void publishResults(CharSequence constraint, FilterResults results) {
            adapter.mFilteredDataset.clear();
            adapter.mDataset.addAll((ArrayList<DataObject>) results.values);
            adapter.notifyDataSetChanged();
        }
    }
}

File: app/src/main/java/com/arukai/uajpspeak/adapter/MyRecyclerViewAdapter.kt
package com.arukai.uajpspeak.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.arukai.uajpspeak.R
import com.arukai.uajpspeak.activity.HomeFragment
import com.arukai.uajpspeak.model.Abecadlo
import com.arukai.uajpspeak.model.DataObject
import com.arukai.uajpspeak.util.LocaleHelper
import com.arukai.uajpspeak.App

class MyRecyclerViewAdapter(
    private val mDataset: ArrayList<DataObject>
) : RecyclerView.Adapter<MyRecyclerViewAdapter.DataObjectHolder>(), Filterable {

    var mFilteredDataset: ArrayList<DataObject> = mDataset
    private var userFilter: UserFilter? = null
    private var myClickListener: MyClickListener? = null

    companion object {
        private const val LOG_TAG = "MyRecyclerViewAdapter"
        private val abc = Abecadlo()
    }

    interface MyClickListener {
        fun onItemClick(position: Int, v: View)
    }

    inner class DataObjectHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        val firstRow: TextView = itemView.findViewById(R.id.textView)
        val secondRow: TextView = itemView.findViewById(R.id.textView2)
        val thirdRow: TextView = itemView.findViewById(R.id.textView3)

        init {
            itemView.setOnClickListener(this)
        }

        override fun onClick(v: View) {
            v.setOnClickListener(null)
            myClickListener?.onItemClick(adapterPosition, v)
        }
    }

    fun setOnItemClickListener(listener: MyClickListener) {
        this.myClickListener = listener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DataObjectHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.card_view_row, parent, false)
        return DataObjectHolder(view)
    }

    override fun onBindViewHolder(holder: DataObjectHolder, position: Int) {
        val part1 = mDataset[position].mText1
        val part2 = mDataset[position].mText2
        val part3 = mDataset[position].mText3

        // The dataset is already filtered by HomeFragment according to current gender.
        // Bind values directly.
        val firstLine = part2
        val ukr = part3.replace("*", "")
        val lang = LocaleHelper.getSavedLanguage(App.appContext)
        val phonetic = if (lang == "ja") abc.convert(part3) else abc.romanize(part3)

        holder.firstRow.text = firstLine
        holder.secondRow.text = ukr
        holder.thirdRow.text = phonetic

        val firstFlagRes = when (lang) {
            "en" -> R.drawable.uk
            "de" -> R.drawable.de
            "fr" -> R.drawable.fr
            "ja" -> R.drawable.jp
            else -> R.drawable.uk
        }
        holder.firstRow.setCompoundDrawablesWithIntrinsicBounds(firstFlagRes, 0, 0, 0)
        holder.secondRow.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ua, 0, 0, 0)
    }

    fun addItem(dataObj: DataObject, index: Int) {
        mDataset.add(index, dataObj)
        notifyItemInserted(index)
    }

    fun deleteItem(index: Int) {
        mDataset.removeAt(index)
        notifyItemRemoved(index)
    }

    override fun getItemCount(): Int = mDataset.size

    override fun getFilter(): Filter {
        if (userFilter == null)
            userFilter = UserFilter(this, mDataset)
        return userFilter!!
    }

    private class UserFilter(
        private val adapter: MyRecyclerViewAdapter,
        originalList: ArrayList<DataObject>
    ) : Filter() {

        private val originalList: ArrayList<DataObject> = ArrayList(originalList)
        private val filteredList: ArrayList<DataObject> = ArrayList()

        override fun performFiltering(constraint: CharSequence): FilterResults {
            filteredList.clear()
            val results = FilterResults()

            if (constraint.isEmpty()) {
                filteredList.addAll(originalList)
            } else {
                val filterPattern = constraint.toString().lowercase().trim()

                for (user in originalList) {
                    val sourceText = user.mText2.lowercase()
                    val ukr = user.mText3.lowercase().replace("*", "")
                    if (sourceText.contains(filterPattern) || ukr.contains(filterPattern)) {
                        filteredList.add(user)
                    }
                }
            }
            results.values = filteredList
            results.count = filteredList.size
            return results
        }

        override fun publishResults(constraint: CharSequence, results: FilterResults) {
            adapter.mFilteredDataset.clear()
            @Suppress("UNCHECKED_CAST")
            adapter.mDataset.addAll(results.values as ArrayList<DataObject>)
            adapter.notifyDataSetChanged()
        }
    }
}
